### ISSUE TITLE
Pylance looses type when inside of for in

### ISSUE THREAD

**Issue by Akindin** [microsoft/pyright#9017]
## Environment data

-   Pylance version: v2024.9.1
-   OS and version: Windows 10 Home 22H2
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.12.4

## Code Snippet

```python
import json
from typing import Literal


ActionType = Literal["keyboard", "mouse"]

class InputSystemInterface:
    def handle(self, request_json: str, d: ActionType) -> str:
        request = json.loads(request_json)
        data = request.get("data")
        a: ActionType  # recognises as type
        try:
            actions: list[dict] = data.get("actions")
            d: ActionType # recognises as type
            for action in actions:
                action_type: ActionType = action[
                    "type"
                ]  # Variable not allowed in type expression (sees it as (variable) ActionType: Any)
                if action_type not in ActionType.__args__:
                    raise ValueError(f"{action_type} is not in {ActionType.__args__}")

        except Exception as error:
            status = "error"
            response_data = {"message": str(error)}

        response = {
            "sender": "InputSystem",
            "recipient": request.get("sender"),
            "status": status,
            "data": response_data,
        }
        return json.dumps(response)


print(
    InputSystemInterface().handle(
        json.dumps(
            {
                "sender": "ClientModule",
                "recipient": "ImageAnalyzer",
                "action": "analyze_image",
                "data": {"actions": [{"type": "432"}]},
            }
        )
    )
)

```

## Repro Steps

1. Try to get typehints for ActionType inside of "for action in action"

## Expected behavior

It resolves ActionType as (type) ActionType = Literal['keyboard', 'mouse']

## Actual behavior

It resolves ActionType as (variable) ActionType: Any

## Logs

```
[{
	"resource": "/C:/Users/<removed for privacy reasons>/src/input_system/__init__.py",
	"owner": "python",
	"code": {
		"value": "reportInvalidTypeForm",
		"target": {
			"$mid": 1,
			"path": "/microsoft/pyright/blob/main/docs/configuration.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "reportInvalidTypeForm"
		}
	},
	"severity": 4,
	"message": "Variable not allowed in type expression",
	"source": "Pylance",
	"startLineNumber": 16,
	"startColumn": 30,
	"endLineNumber": 16,
	"endColumn": 40
}]
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.381 and will be included in a future release of pylance.

**Comment by Akindin** [comment]
@rchiodo actually this 
```python
type ActionType = Literal["keyboard", "mouse"]
```
differs from
```python
ActionType = Literal["keyboard", "mouse"]
```
cause I can't get ActionType.__args__ in first case and I need them to do runtime checks to properly extract error
